src/neon/stage1.rs

```diff
@@ -0,0 +1,531 @@
+    whitespace: &mut u64,
+    structurals: &mut u64,
+) {
+    let low_nibble_mask: uint8x16_t = uint8x16_t::new(16, 0, 0, 0, 0, 0, 0, 0, 0, 8, 12, 1, 2, 9, 0, 0);
+    let high_nibble_mask: uint8x16_t = uint8x16_t::new(8, 0, 18, 4, 0, 1, 0, 1, 0, 0, 0, 3, 2, 1, 0, 0);
+    let structural_shufti_mask: uint8x16_t = vmovq_n_u8(0x7);
+    let whitespace_shufti_mask: uint8x16_t = vmovq_n_u8(0x18);
+    let low_nib_and_mask: uint8x16_t = vmovq_n_u8(0xf);
+
+    let nib_0_lo: uint8x16_t = vandq_u8(input.v0, low_nib_and_mask);
+    let nib_0_hi: uint8x16_t = vshrq_n_u8(input.v0, 4);
+    let shuf_0_lo: uint8x16_t = vqtbl1q_u8(low_nibble_mask, nib_0_lo);
+    let shuf_0_hi: uint8x16_t = vqtbl1q_u8(high_nibble_mask, nib_0_hi);
+    let v_0: uint8x16_t = vandq_u8(shuf_0_lo, shuf_0_hi);
+
+    let nib_1_lo: uint8x16_t = vandq_u8(input.v1, low_nib_and_mask);
+    let nib_1_hi: uint8x16_t = vshrq_n_u8(input.v1, 4);
+    let shuf_1_lo: uint8x16_t = vqtbl1q_u8(low_nibble_mask, nib_1_lo);
+    let shuf_1_hi: uint8x16_t = vqtbl1q_u8(high_nibble_mask, nib_1_hi);
+    let v_1: uint8x16_t = vandq_u8(shuf_1_lo, shuf_1_hi);
+
+    let nib_2_lo: uint8x16_t = vandq_u8(input.v2, low_nib_and_mask);
+    let nib_2_hi: uint8x16_t = vshrq_n_u8(input.v2, 4);
+    let shuf_2_lo: uint8x16_t = vqtbl1q_u8(low_nibble_mask, nib_2_lo);
+    let shuf_2_hi: uint8x16_t = vqtbl1q_u8(high_nibble_mask, nib_2_hi);
+    let v_2: uint8x16_t = vandq_u8(shuf_2_lo, shuf_2_hi);
+
+    let nib_3_lo: uint8x16_t = vandq_u8(input.v3, low_nib_and_mask);
+    let nib_3_hi: uint8x16_t = vshrq_n_u8(input.v3, 4);
+    let shuf_3_lo: uint8x16_t = vqtbl1q_u8(low_nibble_mask, nib_3_lo);
+    let shuf_3_hi: uint8x16_t = vqtbl1q_u8(high_nibble_mask, nib_3_hi);
+    let v_3: uint8x16_t = vandq_u8(shuf_3_lo, shuf_3_hi);
+
+    let tmp_0: uint8x16_t = vtstq_u8(v_0, structural_shufti_mask);
+    let tmp_1: uint8x16_t = vtstq_u8(v_1, structural_shufti_mask);
+    let tmp_2: uint8x16_t = vtstq_u8(v_2, structural_shufti_mask);
+    let tmp_3: uint8x16_t = vtstq_u8(v_3, structural_shufti_mask);
+    *structurals = neon_movemask_bulk(tmp_0, tmp_1, tmp_2, tmp_3);
+
+    let tmp_ws_0: uint8x16_t = vtstq_u8(v_0, whitespace_shufti_mask);
+    let tmp_ws_1: uint8x16_t = vtstq_u8(v_1, whitespace_shufti_mask);
+    let tmp_ws_2: uint8x16_t = vtstq_u8(v_2, whitespace_shufti_mask);
+    let tmp_ws_3: uint8x16_t = vtstq_u8(v_3, whitespace_shufti_mask);
+    *whitespace = neon_movemask_bulk(tmp_ws_0, tmp_ws_1, tmp_ws_2, tmp_ws_3);
+}
+
+// flatten out values in 'bits' assuming that they are are to have values of idx
+// plus their position in the bitvector, and store these indexes at
+// base_ptr[base] incrementing base as we go
+// will potentially store extra values beyond end of valid bits, so base_ptr
+// needs to be large enough to handle this
+//TODO: usize was u32 here does this matter?
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn flatten_bits(base: &mut Vec<u32>, idx: u32, mut bits: u64) {
```

Comment:
The upstream code has gone to unroll this loop, I'll open a ticket (#37) to adopt that in a general ways  but since this code is still getting worked on we could start with it here?

Reply:
On looking at this, we have a slightly different appraoch, it might be worth comparing the two. and postpone this change for later after doing some benchmarks. Our implementation might be faster as it stands but could be even faster faster w/ unrolling. Feel free to discard if you feel like it doesn't belong here.

Reply:
I'd love to bring it in from upstream soon as part of some arch consolidation work!